File: app/src/main/java/com/apppoweron/circularrevealbuttondemo/BaseActivity.java
package com.apppoweron.circularrevealbuttondemo;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;

public abstract class BaseActivity extends AppCompatActivity implements FragmentCommunicator {

    private static final int DEFAULT_CONTAINER_ID = -1;

    private BackPressListener mBackPressListener;
    private FragmentTransaction mDaleayedTransaction;
    private boolean mIsSavedInstanceStateCalled;

    private void addFragment(Fragment fragment) throws NoContainerException {
        if (!hasContainerId()) {
            throw new NoContainerException();
        }
        getSupportFragmentManager().beginTransaction().add(getContainerId(), fragment, fragment.getClass().getSimpleName() + "_TAG").commit();
    }

    private void replaceFragment(Fragment fragment, boolean needToBackStack) throws NoContainerException {
        if (!hasContainerId()) {
            throw new NoContainerException();
        }
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction().replace(getContainerId(), fragment, fragment.getClass().getSimpleName() + "_TAG");
        if (needToBackStack) {
            transaction.addToBackStack(null);
        }
        if (mIsSavedInstanceStateCalled) {
            mDaleayedTransaction = transaction;
            return;
        }
        transaction.commit();

        if (needToBackStack) {
            switchBackButton(true);
        }
    }

    public void loadFragment(Fragment fragment, FragmentLoadType loadType) throws NoContainerException {
        loadFragment(fragment, loadType, false);
    }


    public void loadFragment(Fragment fragment, FragmentLoadType loadType, boolean needToBackStack) throws NoContainerException {
        switch (loadType) {
            case ADD:
                addFragment(fragment);
                break;
            case REPLACE:
                replaceFragment(fragment, needToBackStack);
                break;
        }
    }

    private boolean hasContainerId() {
        return getContainerId() != DEFAULT_CONTAINER_ID;
    }

    protected int getContainerId() {
        return 0;
    }

    @Override
    public void onBackPressed() {
        if (getSupportFragmentManager().getBackStackEntryCount() == 1) {
            switchBackButton(false);
        }
        if (mBackPressListener != null) {
            mBackPressListener.onBackPressed();
        } else {
            callSuperBackPressed();
        }
    }

    public void setBackPressListener(BackPressListener listener) {
        mBackPressListener = listener;
    }

    public void callSuperBackPressed() {
        super.onBackPressed();
    }

    interface BackPressListener {
        void onBackPressed();
    }

    public enum FragmentLoadType {
        ADD, REPLACE
    }


    /**
     * Remove items from backstack
     */
    protected void clearBackStack() {
        getSupportFragmentManager().popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
    }

    @Override
    public void onNewFragmentSelected(Fragment fragment, boolean replaceIt, boolean needToBackStack) throws NoContainerException {
        if (replaceIt) {
            loadFragment(fragment, FragmentLoadType.REPLACE, needToBackStack);
        } else {
            loadFragment(fragment, FragmentLoadType.ADD, needToBackStack);
        }

    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        mIsSavedInstanceStateCalled = true;
        super.onSaveInstanceState(outState);
    }

    @Override
    protected void onResume() {
        super.onResume();
        mIsSavedInstanceStateCalled = false;
        if (mDaleayedTransaction != null) {
            mDaleayedTransaction.commit();
            mDaleayedTransaction = null;
        }
    }

    /**
     * Show backbutton on toolbar
     *
     * @param turnOn show it or not
     */
    private void switchBackButton(boolean turnOn) {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(turnOn);
            getSupportActionBar().setHomeButtonEnabled(turnOn);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }

    }

}

File: app/src/main/kotlin/com/apppoweron/circularrevealbuttondemo/BaseActivity.kt
package com.apppoweron.circularrevealbuttondemo

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentTransaction
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem

abstract class BaseActivity : AppCompatActivity(), FragmentCommunicator {

    companion object {
        private const val DEFAULT_CONTAINER_ID = -1
    }

    private var mBackPressListener: BackPressListener? = null
    private var mDelayedTransaction: FragmentTransaction? = null
    private var mIsSavedInstanceStateCalled: Boolean = false

    protected open val containerId: Int
        get() = 0

    @Throws(NoContainerException::class)
    private fun addFragment(fragment: Fragment) {
        if (!hasContainerId()) {
            throw NoContainerException()
        }
        supportFragmentManager.beginTransaction().add(containerId, fragment, fragment.javaClass.getSimpleName() + "_TAG").commit()
    }

    @Throws(NoContainerException::class)
    private fun replaceFragment(fragment: Fragment, needToBackStack: Boolean) {
        if (!hasContainerId()) {
            throw NoContainerException()
        }
        val transaction = supportFragmentManager.beginTransaction().replace(containerId, fragment, fragment.javaClass.getSimpleName() + "_TAG")
        if (needToBackStack) {
            transaction.addToBackStack(null)
        }
        if (mIsSavedInstanceStateCalled) {
            mDelayedTransaction = transaction
            return
        }
        transaction.commit()

        if (needToBackStack) {
            switchBackButton(true)
        }
    }


    @Throws(NoContainerException::class)
    @JvmOverloads
    fun loadFragment(fragment: Fragment, loadType: FragmentLoadType, needToBackStack: Boolean = false) {
        when (loadType) {
            BaseActivity.FragmentLoadType.ADD -> addFragment(fragment)
            BaseActivity.FragmentLoadType.REPLACE -> replaceFragment(fragment, needToBackStack)
        }
    }

    private fun hasContainerId(): Boolean {
        return containerId != DEFAULT_CONTAINER_ID
    }

    override fun onBackPressed() {
        if (supportFragmentManager.backStackEntryCount == 1) {
            switchBackButton(false)
        }
        if (mBackPressListener != null) {
            mBackPressListener!!.onBackPressed()
        } else {
            callSuperBackPressed()
        }
    }

    internal fun setBackPressListener(listener: BackPressListener) {
        mBackPressListener = listener
    }

    private fun callSuperBackPressed() {
        super.onBackPressed()
    }

    internal interface BackPressListener {
        fun onBackPressed()
    }

    enum class FragmentLoadType {
        ADD, REPLACE
    }


    /**
     * Remove items from backstack
     */
    protected fun clearBackStack() {
        supportFragmentManager.popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
    }

    @Throws(NoContainerException::class)
    override fun onNewFragmentSelected(fragment: Fragment, replaceIt: Boolean, needToBackStack: Boolean) {
        if (replaceIt) {
            loadFragment(fragment, FragmentLoadType.REPLACE, needToBackStack)
        } else {
            loadFragment(fragment, FragmentLoadType.ADD, needToBackStack)
        }

    }

    override fun onSaveInstanceState(outState: Bundle) {
        mIsSavedInstanceStateCalled = true
        super.onSaveInstanceState(outState)
    }

    override fun onResume() {
        super.onResume()
        mIsSavedInstanceStateCalled = false
        if (mDelayedTransaction != null) {
            mDelayedTransaction!!.commit()
            mDelayedTransaction = null
        }
    }

    /**
     * Show backbutton on toolbar
     *
     * @param turnOn show it or not
     */
    private fun switchBackButton(turnOn: Boolean) {
        if (supportActionBar != null) {
            supportActionBar!!.setDisplayHomeAsUpEnabled(turnOn)
            supportActionBar!!.setHomeButtonEnabled(turnOn)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }

    }



}
